File: src/main/java/be/vsol/vsol6/model/organization/Study.java
package be.vsol.vsol6.model.organization;

import be.vsol.database.annotations.db;
import be.vsol.tools.json;
import be.vsol.vsol4.Vsol4Study;
import be.vsol.vsol6.model.Record;

import java.time.LocalDateTime;
import java.util.Collections;
import java.util.Comparator;

public class Study extends Record {

    @json @db private String description;
    @json @db private LocalDateTime dateTime;
    @json @db private int seriesCount;
    @json private Patient patient;

    // Constructors

    public Study() { }

    public Study(Vsol4Study vsol4Study) {
        super(vsol4Study.getId(), vsol4Study.getLastOpenedDate());

        this.description = vsol4Study.getDescription();
        this.dateTime = vsol4Study.getDate();
        this.seriesCount = vsol4Study.getSeriesCount();
        this.patient = new Patient(vsol4Study.getPatient());
        this.patient.setClient(new Client(vsol4Study.getClient()));
    }

    // Methods

    public Vsol4Study getVsol4Study() {
        return new Vsol4Study(id, description, dateTime, patient == null ? null : patient.getVsol4Patient(), seriesCount);
    }

    public String getClientString() {
        return patient.getClient().toString();
    }

    public String getPatientString() {
        return patient.toString();
    }

    @Override public String toString() {
        return description;
    }

    // Static Methods

    public static Comparator<Study> getComparator(String sortField, boolean sortAsc) {
        Comparator<Study> result = switch (sortField) {
            case "client" -> Comparator.comparing(Study::getClientString);
            case "patient" -> Comparator.comparing(Study::getPatientString);
            case "date" -> Comparator.comparing(Study::getDateTime);
            case "description" -> Comparator.comparing(Study::getDescription);
            case "seriesCount" -> Comparator.comparingInt(Study::getSeriesCount);
            default -> null;
        };

        if (result == null) {
            return Collections.reverseOrder(Comparator.comparing(Study::getLastOpenedTime));
        } else {
            return sortAsc ? result : Collections.reverseOrder(result);
        }
    }

    // Getters

    public String getDescription() { return description; }

    public LocalDateTime getDateTime() { return dateTime; }

    public int getSeriesCount() { return seriesCount; }

    public Patient getPatient() { return patient; }

    // Setters

    public void setDescription(String description) { this.description = description; }

    public void setDateTime(LocalDateTime dateTime) { this.dateTime = dateTime; }

    public void setSeriesCount(int seriesCount) { this.seriesCount = seriesCount; }

    public void setPatient(Patient patient) { this.patient = patient; }

}

File: src/main/resources/js/model/Study.js
import { VsolRecord } from "./VsolRecord.js";
import { Patient } from "./Patient.js";
export class Study extends VsolRecord {
    constructor(id = null, patient = new Patient(), dateTime = new Date(), description = "", seriesCount = 0) {
        super(id);
        this.patient = patient;
        this.dateTime = dateTime;
        this.description = description;
        this.seriesCount = seriesCount;
    }
    static from(src) {
        return src == null ? null : new Study(src.id, Patient.from(src.patient), src.dateTime, src.description, src.seriesCount);
    }
    static fromRows(rows) {
        let result = [];
        for (let row of rows) {
            result.push(Study.from(row));
        }
        return result;
    }
    // Getters
    getPatient() { return this.patient; }
    getClient() { return this.patient == null ? null : this.patient.getClient(); }
    getDateTime() { return this.dateTime; }
    getDescription() { return this.description; }
    getSeriesCount() { return this.seriesCount; }
}
